torch/distributed/_composable/replicate.py

```diff
@@ -0,0 +1,96 @@
+import torch
+import torch.nn as nn
+from . import _ddp
+
+from typing import List, Tuple
+
+
+class DistributedState:
+    ...
+
+
+class ReplicateState(DistributedState):
+    def __init__(self) -> None:
+        self.modules: List[nn.Module] = []
+        self.has_initialized: bool = False
+        self._param_list: nn.ParameterList = nn.ParameterList()
+
+    def mark_modules(self, *modules: nn.Module) -> None:
+        for module in modules:
+            self.modules.append(module)
+            # TODO(@yhcharles): do not directly set attribute
+            module._distributed_state = self  # type: ignore[assignment]
+
+    def _recursive_collect_params(self, module: nn.Module) -> None:
+        if (
+            getattr(module, "_distributed_state", None) is not None
```

Comment:
also seems module._distributed_state is not None for modules that are marked as replicated at this point, so it always returns and did not fill in 'self._param_list'?

Reply:
Updated